src/borda_clj/client.clj

```diff
@@ -16,30 +17,57 @@
+(comment
+  (defn reducing-submitter
+    "Returns two functions. The first is a reducing submitter that collects and
    aggregates measurements and sends them using the given send function at the
    specified interval (in milliseconds) and limits the size of pending
    measurements to max-buffer-size. The second is a function that flushes the
    buffer and stops the background thread that does the submitting."
+    [max-buffer-size interval send on-send-error]
+    (let [measurements  (atom {})
+          running       (atom true)
+          submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
+          resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
+          flush         (fn [on-flush-error]
+                          (let [[m] (reset-vals! measurements {})]
+                            (if (not-empty m)
+                              (try
+                                (send m)
+                                (catch Throwable e
+                                  (on-flush-error m e))))))
+          stop          (fn [] (reset! running false) (flush on-send-error))]
+                                        ; periodically send to borda
+      (future (while @running (do
+                                (Thread/sleep interval)
+                                (flush (fn [m e] (on-send-error m e) (resubmit m))))))
+      [submit stop])))
+
+(defn reducing-submitter
   [max-buffer-size interval send on-send-error]
-  (let [measurements  (atom {})
-        running       (atom true)
-        submit        (fn [dimensions values] (swap! measurements collect max-buffer-size dimensions values))
-        resubmit      (fn [m] (doseq [[dimensions values] m] (submit dimensions values)))
-        flush         (fn [on-flush-error]
-                        (let [[m] (reset-vals! measurements {})]
-                          (if (not-empty m)
-                            (try
-                              (send m)
-                              (catch Throwable e
-                                (on-flush-error m e))))))
-        stop          (fn [] (reset! running false) (flush on-send-error))]
-        ; periodically send to borda
-    (future (while @running (do
-                              (Thread/sleep interval)
-                              (flush (fn [m e] (on-send-error m e) (resubmit m))))))
-    [submit stop]))
+  (let [report-ch (chan)]
+    (go-loop [measurements {}
+              flush-timeout (timeout interval)]
+      (let [flush (fn [on-error]
+                    (when (not-empty measurements)
+                      (try
+                        (send measurements)
+                        (catch Throwable e
+                          (on-error measurements e)))))
+            [[dimensions values :as v] ch] (alts! [report-ch flush-timeout])]
+          (cond (= ch flush-timeout)
+                (do
+                  (flush (fn [m e]
```

Comment:
I didn't want to change any functionality, but I'm not convinced it's worth exposing this hook at all, and even if we do, I'm not convinced that adding the measurements in this callback is of much use.  The Pro Server isn't using the measurements in its handler, for example.

Reply:
Sorry, I don't quite understand. What hook are you talking about?

Reply:
`on-send-error`.  The client of `reducing-submitter` is providing the sender function to begin with, so they can handle errors there themselves if they want to do anything beyond logging them.

Reply:
Oh, I see what you mean. The main use for that `on-error` callback is to facilitate logging, and I assumed that sometimes the caller might want to log which measurements (or the number of measurements) that failed to submit (like we do in the unit test actually).